onnxruntime/core/providers/cuda/controlflow/loop.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/controlflow/loop.h"
+#include "core/providers/cuda/cuda_common.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace onnxruntime::common;
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(Loop,
+                                  kOnnxDomain,
+                                  1, 10,
+                                  kCudaExecutionProvider,
+                                  KernelDefBuilder()
+                                      .InputMemoryType<OrtMemTypeCPUInput>(0)  // 'M' needs to be on CPU
+                                      .InputMemoryType<OrtMemTypeCPUInput>(1)  // 'cond' needs to be on CPU
+                                      .TypeConstraint("I", DataTypeImpl::GetTensorType<int64_t>())
+                                      .TypeConstraint("B", DataTypeImpl::GetTensorType<bool>())
+                                      .TypeConstraint("V", DataTypeImpl::AllFixedSizeTensorTypes()),
+                                  Loop);
+
+// zero variadic argument support was added in opset 11. using same implementation as for previous version
+ONNX_OPERATOR_KERNEL_EX(Loop,
+                        kOnnxDomain,
+                        11,
+                        kCudaExecutionProvider,
+                        KernelDefBuilder()
+                            .InputMemoryType<OrtMemTypeCPUInput>(0)  // 'M' needs to be on CPU
+                            .InputMemoryType<OrtMemTypeCPUInput>(1)  // 'cond' needs to be on CPU
+                            .TypeConstraint("I", DataTypeImpl::GetTensorType<int64_t>())
+                            .TypeConstraint("B", DataTypeImpl::GetTensorType<bool>())
+                            .TypeConstraint("V", DataTypeImpl::AllFixedSizeTensorTypes()),
+                        Loop);
+
+static Status ConcatenateGpuOutput(std::vector<OrtValue>& per_iteration_output,
+                                   void* output, size_t output_size_in_bytes) {
+  const auto& first_output = per_iteration_output.front().Get<Tensor>();
+  const auto& per_iteration_shape = first_output.Shape();
+  size_t bytes_per_iteration = first_output.SizeInBytes();
+
+  void* cur_output = output;
+  for (size_t i = 0, num_iterations = per_iteration_output.size(); i < num_iterations; ++i) {
+    auto& ort_value = per_iteration_output[i];
+    auto& iteration_data = ort_value.Get<Tensor>();
+
+    // sanity check
```

Comment:
Do we need to check the type as well ?

Reply:
I don't believe so. Type checks are done in the type/shape inferencing and this is just a byte copy of the output. We know each loop produces the same type each time, and we know that matches the Loop output type.

---
In reply to: [351567332](https://github.com/microsoft/onnxruntime/pull/2444#discussion_r351567332) [](ancestors = 351567332)